### ISSUE TITLE
Libnd4j: Add Dynamic Partition Backprop

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#6054]
Backprop: it's basically dynamic stich, but we need to create the index array from the partitions array.
So if we have partition indices [0, 0, 1, 1, 0, 1] then we want to infer indices [0,1,4] and [2,3,5]. Then it's standard dynamic stitch on the gradient arrays.

I'd implement it in SameDiff/Java, but can't work out how to get the indices (I'd use this if we had it: https://github.com/deeplearning4j/deeplearning4j/issues/6053 - or an "index of values matching condition" op)

https://www.tensorflow.org/api_docs/python/tf/dynamic_partition
https://www.tensorflow.org/api_docs/python/tf/dynamic_stitch

**Comment by shugeo** [comment]
This op is more similar to dynamic_partition than dynamic_stitch.
